### DISCUSSION TITLE
Why is "Mute Myself" still sending data?

### DISCUSSION THREAD

**Discussion by 38github** [jamulussoftware/jamulus#1160] [General]
I would just like to ask why "mute myself" is sending full stream data to the server? Is it possible to minimize this? Would this be possible with variable bitrate or is opus lack of "quality setting" making this problematic? With quality setting I am refering to how ogg vorbis' -q and LAME's -V commands do not try to follow a defined bitrate but instead quality.

**Comment by 38github** [comment]
I am sorry for posting this in Issues. Please move this thread to Discussion, please.

**Reply by ann0see** [reply]
Done.

**Comment by dcorson-ticino-com** [comment]
With mute myself all the others can't hear you, but you can hear yourself, through the server.
This is so you can test the quality of the sound channel, do setup, tuning, without disturbing others.

**Comment by 38github** [comment]
> With mute myself all the others can't hear you, but you can hear yourself, through the server.
> This is so you can test the quality of the sound channel, do setup, tuning, without disturbing others.

I know but why does Jamulus use network bandwidth when the sound is local?

**Reply by gilgongo** [reply]
At the risk of asking a stupid question, if you are successfully sending ~900kbps to the server when you are **not** muted, why would it be a problem when you are muted? If there's not enough bandwidth to play, then that's a separate issue, isn't it?

**Reply by 38github** [reply]
I am just thinking that it is a waste of bandwidth.

**Reply by gilgongo** [reply]
I see. There has been some discussion of implementing a "listening mode" for people who are not intending to send sound to the server (to prevent servers for large ensembles using too much bandwidth I think), but the consensus on that was to use a secondary method like Zoom (or the new streamer). Bandwidth isn't generally seen as a problem compared to other issues that tend to arise as far as I know.

**Comment by 38github** [comment]
> through the server.

But is it really? That would mean that I would get the same latency using Mute Myself as I do when I am not and that is not the case. Isn't it a waste of bandwidth to send ~900kbps to the server even though I don't want to transmit anything to anyone?

**Comment by chrisrimple** [comment]
EDIT: I misunderstood the question.  As noted by others, when Mute Myself is enabled, the sound is heard locally but "silence" is sent to the server.

**Comment by 38github** [comment]
Either I am not explaining myself well or you are missing something. When I have Mute Myself the sound can not possibly be sent to the server and back. I have connected to a server half way around the globe and with Mute Myself disabled I get at least 250ms delay. With Mute Myself enabled the sound is immediate. Therefore the sound can not be sent to the server and back. Therefore, why is ~900kbps still being sent to the server?

**Reply by dcorson-ticino-com** [reply]
If this is truly the case, then I wonder why we don't already have our setup loop-back mode.
This is not what I understood from Volker's explanations, but maybe it was me misunderstanding.
Someone who has some visibility in the sound data routing should get on this right away, we have had so many requests for this..

**Reply by ann0see** [reply]
> Someone who has some visibility in the sound data routing should get on this right away, we have had so many requests for this..

I totally agree! It would also involve some UI changes probably (sliders in "off" state) which might be confusing (there is a disagreement between some people since it should be clear that the mix is on the server).

What could/should be done here:

* Don't show the mixer board in off state, just the input level meters and give back sound. 
* Give error messages if e.g. the sample rate is wrong or a wrong ASIO driver was chosen, but don't raise fatal error messages which close Jamulus this might solve multiple usablity issues but requires some re writing of the audio code.

**Reply by dakhubgit** [reply]
> If this is truly the case, then I wonder why we don't already have our setup loop-back mode.
> This is not what I understood from Volker's explanations, but maybe it was me misunderstanding.

It is the case.  Zeros are sent and the audio is looped locally.  I've seen the code.  Part of the problem I think is simply that the server is only told stereosity and quality of the signal it is supposed to send, and the assumption is that this is what it will receive.  So not just does "Mute self" send a zero signal, but also "Mono-In/Stereo-Out" sends two channels of identical signal.  Combine the two, and you get two channels of zeros sent, encoded at fixed bit rate.

**Comment by ann0see** [comment]
> but you can hear yourself, through the server.

As far as I know, that's **not** true. Jamulus just sends silence to stay connected to the server.

**Comment by 38github** [comment]
@ann0see Thank you for replying! I have a couple of questions if you have nothing against answering them.

- Question #1: Is Jamulus using CBR and not VBR?

I know opus is bad at using less bitrate than it needs. If I encode a sine tone with opus VBR at 512kbps and 2,5ms frame length I get almost 512kbps. If I encode music with a lot of frequencies it sometimes uses even less than the sine tone sample.

- Question #2: If it is CBR, would having it use VBR create any issues?

- Question #3: Why is the sound quality of Jamulus using 64 samples, enable low network buffer, high quality and stereo so bad compared to encoding an opus file with the settings I mentioned earlier?

I especially hear it with a distorted guitar tone slightly panned to the side with a roomy reverb.

**Reply by ann0see** [reply]
I'm afraid I can't answer these questions since I haven't looked much into the actual audio processing code (Furthermore I'm not a C++ dev and was mainly working on the installer & only looked into the actual code because of my implementation of the waiting room feature). 
I only "know" the answer from above since dakhubgit mentioned the code somewhere else and a real connection handshake or similar doesn't seem to exist.

To prevent sending silence, maybe some TCP magic or protocol messages would have to be implemented. But that's an area I don't have too much experience with ;-).

**Reply by 38github** [reply]
I really appreciate you taking the time though!

**Reply by ann0see** [reply]
No problem! I'd suggest you to read through the code especially the server.cpp file, I think. Maybe this answers your questions.

**Reply by 38github** [reply]
It seems it is using CBR (row 286, server.cpp):
`// we require a constant bit rate`

**Reply by hoffie** [reply]
> Question #1: Is Jamulus using CBR and not VBR?

As you already found, Jamulus is CBR-only. :)

> Question #2: If it is CBR, would having it use VBR create any issues?

I did a real quick test some weeks ago and the short answer is: Yes, simply changing the codec parameters to VBR will cause Jamulus to break badly. One reason for this is that several code paths (mainly: protocol and jitter buffer) rely on constant frame sizes.
I do not know about any other reasons for not using VBR. I was planning to look into this some more in order to find out if/what/how much would have to be changed if one wanted to use VBR.

You might also be interested in the following issues:
https://github.com/jamulussoftware/jamulus/issues/842
https://github.com/jamulussoftware/jamulus/issues/270#issuecomment-634021946

**Reply by 38github** [reply]
Thank you very much and I will also read those threads you suggested.

**Reply by dakhubgit** [reply]
> @ann0see Thank you for replying! I have a couple of questions if you have nothing against answering them.
> 
>     * Question #1: Is Jamulus using CBR and not VBR?

Yes.
 
> I know opus is bad at using less bitrate than it needs. If I encode a sine tone with opus VBR at 512kbps and 2,5ms frame length I get almost 512kbps. If I encode music with a lot of frequencies it sometimes uses even less than the sine tone sample.
> 
>     * Question #2: If it is CBR, would having it use VBR create any issues?

VBR takes more time to encode since it weeds out the CBR coefficients.  At 64bit sample size, there is not a lot potential for saving.

>     * Question #3: Why is the sound quality of Jamulus using 64 samples, enable low network buffer, high quality and stereo so bad compared to encoding an opus file with the settings I mentioned earlier?

Opus files tend to encode 20ms.  That gives a lot more potential for utilising redundancies particularly for lower frequencies.  The lowest packet size of an Opus file is 2.5ms, namely 120 samples.  Personally, I think that Jamulus should use standard (rather than custom) encoding and bite the bullet to have the smallest size 120 samples.  The encoders are optimised for it, the encoded data can just be dumped as-is to a file (after wrapping into OGG) for recording purposes, and CPU timers actually can do 2.5ms rather than 4/3 ms or 8/3 ms.

> I especially hear it with a distorted guitar tone slightly panned to the side with a roomy reverb.

Accordion tremolo breaks MP3 unless you use ridiculously high bit rates.  Either Vorbis or Opus fare a lot better.  Probably a lot of encoders have some "nemesis".

**Comment by gene96817** [comment]
This is an example of a topic where I wish there was a diagram of the audio processing. Then it would be easier to communicate where the "mute myself" is functioning and how the audio processing is being affected.

**Reply by gilgongo** [reply]
I think there have been various attempts at that, but each time they've either too high or low level, or just not understood. Somebody was working on one a couple of months ago.

**Reply by gene96817** [reply]
Too high level = for general user

Too low level = requires reading the code 

Just right: 
-- provides the map of the audio stream through the software modules doing the processing
-- shows where a UI control or command line parameters changes the behavior of the software module that produces the results.
-- helps to discover some software functions not in the map.
-- probably exposes where a control makes changes in multiple places in the software (and sometimes we forget what are all the changes need to be done at the same time).

**Reply by gilgongo** [reply]
There's this https://github.com/jamulussoftware/jamulus/issues/64#issuecomment-613002886 and https://github.com/jamulussoftware/jamulus/issues/64#issuecomment-633309575 and one posted here https://github.com/jamulussoftware/jamulus/issues/64#issuecomment-754444792

Are those any help? In my experience, diagrams for software function are hard to do because they are encoded in a "language" that only the creator really understands.

**Reply by ann0see** [reply]
@gene96817 Yes probably we need to re think some of the diagrams. There's also the need of redesigning the pictures with the new logo. @mulyaj sorry for tagging you here. I know you're working on something else, but can you change the pictures for the documentation with the new logos?

**Reply by mulyaj** [reply]
I can do that - is it just that diagram?

**Reply by ann0see** [reply]
Yes. I hope the files are still ok. Once you've finished, just open a PR on the website repo and also include the modified source files. I'd like to translate them in German. BTW. the term Central Server might need to be adapted if they're in this picture.

**Reply by gilgongo** [reply]
Yes, "Central Server" is being replaced by (preferably) just "Directory" so that the word "server" isn't used in two different contexts to be confusing (eg "connect to the server"). But it can be "Directory Server" if you need it to be (think twice about possible confusion if you do that).

**Reply by gene96817** [reply]
Would "Directory (Server)" help as a transition notation?  Going straight to "Directory" is better. Anything else can be for a transition; just as long as we remember to drop the transition for the next release.

**Reply by gilgongo** [reply]
I think a word in brackets would be taken to mean that the two words mean the same thing, which is exactly not what what we want people to think :-) Besides, the number of people concerned with directories is tiny, so we needn't worry about them - it's to disambiguate (and update) the term for the majority.

**Reply by mulyaj** [reply]
I've updated the diagram on [this PR here](https://github.com/jamulussoftware/jamuluswebsite/pull/353).
